### DISCUSSION TITLE
Multiple States

### DISCUSSION THREAD

**Discussion by johnallsup** [MusicPlayerDaemon/MPD#2010] [Ideas]
At present, what I have on my music player (a headless Thinkpad in the corner of my flat connected via an audio interface to my hifi), is 32 instances of mpd running as a user, on ports 6600..6631. Each of them can have its own play state (playlist + what track is playing and where in the file it is, volume, and repeat/single/random/consume state). 

This could be subsumed into a single instance of mpd if it had the facility to maintain multiple states and switch between them (only one mpd instance is playing at a time under normal circumstances). 

Also the capacity to dump and load state to named files (not just playlists), and have multiple loaded states in named/numbered slots. Indeed a stack-like system whereby I load an auxiliary playlist in consume mode and, when empty, it pops off the stack and restores (and starts playing) the previous state.

I'm interested to play around with mpd's source code, and implementing this seems like a nice exercise in learning it. Any pointers as to where to start?

Related, is it possible to tell mpd to refresh the contents of the playlists without rescanning the entire library?

**Comment by MaxKellermann** [comment]
Such a feature already exists, it was added in version 0.22. Each MPD "partition" has its own queue and state.

**Reply by johnallsup** [reply]
Thanks. I'd missed that. Though it doesn't seem there is support for it in mpc and ncmpcpp that I can find. (mpc gives unknown command 'partition' for example)

**Reply by johnallsup** [reply]
An additional question: is it possible to copy state from one partition to another?

**Reply by MaxKellermann** [reply]
There is no MPD command to do this - adding one might be useful.

**Comment by naglis** [comment]
> Though it doesn't seem there is support for it in mpc and ncmpcpp that I can find. (mpc gives unknown command 'partition' for example)

`mpc` has support for partitions, see [Partition Commands](https://github.com/MusicPlayerDaemon/mpc/blob/master/doc/index.rst#partition-commands).

Use `mpc -a <partition>` or `mpc --partition <partition>` to execute commands in non-default partition.

**Reply by MaxKellermann** [reply]
And ncmpc, you can create/manage partitions on the outputs screen ("8").

**Reply by johnallsup** [reply]
Thanks. The mpc -a argument was added in 0.36 (I had been using 0.34 since that's what comes with Ubuntu 22.04 LTS). In the case of ncmpc and ncmpcpp I can't find out to select a partition so that I can edit its playlist, move output to it, and play etc.

**Reply by MaxKellermann** [reply]
ncmpcpp (the confusingly named client) has no partition support. Only ncmpc (the original) has.
